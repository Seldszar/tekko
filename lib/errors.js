/**
 * A base error.
 * @private
 * @extends Error
 * @param {...any} params the parameters
 */
class BaseError extends Error {
  constructor(...params) {
    super(...params);

    /* istanbul ignore next */
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }

    Object.defineProperty(this, "name", {
      value: this.constructor.name,
    });
  }
}

/**
 * A parse error.
 * @extends BaseError
 */
class ParseError extends BaseError {}

/**
 * A format error.
 * @extends BaseError
 */
class FormatError extends BaseError {}

exports.ParseError = ParseError;
exports.FormatError = FormatError;
