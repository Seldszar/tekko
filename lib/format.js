const { FormatError } = require("./errors");
const { escapeString } = require("./helpers");

/**
 * Formats message tags.
 * @private
 * @param {Object} input the message tags
 * @return {String} the formatted message tags.
 */
function formatTags(input) {
  const tagPairs = Object.entries(input);
  const tagPairsLength = tagPairs.length;

  let output = "";

  for (let i = 0; i < tagPairsLength; i += 1) {
    const [key, value] = tagPairs[i];

    output += key;

    if (value !== true) {
      output += `=${escapeString(value)}`;
    }

    if (i < tagPairsLength - 1) {
      output += ";";
    }
  }

  return output;
}

/**
 * Formats a message prefix.
 * @private
 * @throws {FormatError}
 * @param {MessagePrefix} input the message prefix
 * @return {String} the formatted message prefix.
 */
function formatPrefix(input) {
  if (input.name) {
    let output = input.name;

    if (input.host) {
      if (input.user) {
        output += `!${input.user}`;
      }

      output += `@${input.host}`;
    }

    return output;
  }

  throw new FormatError("Invalid prefix");
}

/**
 * Formats a message.
 * @throws {FormatError}
 * @param {Message} input the message
 * @return {?String} the formatted message.
 */
function format(input) {
  if (!input || !input.command) {
    throw new FormatError("Invalid message");
  }

  let output = "";

  if (input.tags && Object.keys(input.tags).length > 0) {
    output += `@${formatTags(input.tags)}`;
  }

  if (input.prefix && Object.keys(input.prefix).length > 0) {
    if (output) {
      output += " ";
    }

    output += `:${formatPrefix(input.prefix)}`;
  }

  if (output) {
    output += " ";
  }

  output += input.command;

  if (input.params) {
    const paramsLength = input.params.length;

    for (let i = 0; i < paramsLength; i += 1) {
      output += ` ${escapeString(input.params[i])}`;
    }
  }

  if (input.trailing) {
    output += ` :${input.trailing}`;
  }

  return output;
}

module.exports = format;
