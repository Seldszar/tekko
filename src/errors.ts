/**
 * A base error.
 * @param message the message
 */
class BaseError extends Error {
  constructor(message: string) {
    super(message);

    /* istanbul ignore next */
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }

    Object.defineProperty(this, "name", {
      value: this.constructor.name,
    });
  }
}

/**
 * A parsing error.
 */
export class ParseError extends BaseError {
}

/**
 * A formatting error.
 */
export class FormatError extends BaseError {
}
