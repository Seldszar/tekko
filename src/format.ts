import { FormatError } from "./errors";
import { escapeString } from "./helpers";
import { MessagePrefix, MessageTags } from "./types";

/**
 * Formats message tags.
 * @param input the message tags
 * @return the formatted message tags.
 */
function formatTags(input: MessageTags): string {
  const tagPairs = Object.entries(input);
  const tagPairsLength = tagPairs.length;

  let output = "";

  for (let i = 0; i < tagPairsLength; i += 1) {
    const [key, value] = tagPairs[i];

    output += key;

    if (value !== true) {
      output += `=${escapeString(value)}`;
    }

    if (i < tagPairsLength - 1) {
      output += ";";
    }
  }

  return output;
}

/**
 * Formats a message prefix.
 * @throws {FormatError}
 * @param input the message prefix
 * @return the formatted message prefix.
 */
function formatPrefix(input: MessagePrefix): string {
  if (input.name) {
    let output = input.name;

    if (input.host) {
      if (input.user) {
        output += `!${input.user}`;
      }

      output += `@${input.host}`;
    }

    return output;
  }

  throw new FormatError("Invalid prefix");
}

/**
 * Formats a message.
 * @throws {FormatError}
 * @param input the message
 * @return the formatted message.
 */
export function format(input: any): string {
  if (!input || !input.command) {
    throw new FormatError("Invalid message");
  }

  let output = "";

  if (input.tags && Object.keys(input.tags).length > 0) {
    output += `@${formatTags(input.tags)}`;
  }

  if (input.prefix && Object.keys(input.prefix).length > 0) {
    if (output) {
      output += " ";
    }

    output += `:${formatPrefix(input.prefix)}`;
  }

  if (output) {
    output += " ";
  }

  output += input.command;

  if (input.params && input.params.length > 0) {
    const trailing = input.params[input.params.length - 1];
    const middle = input.params.slice(0, -1);
    const middleLength = middle.length;

    for (let i = 0; i < middleLength; i += 1) {
      output += ` ${escapeString(middle[i])}`;
    }

    output += ` :${trailing}`;
  }

  return output;
}
