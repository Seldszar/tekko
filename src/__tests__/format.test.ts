import { FormatError } from "../errors";
import { format as formatMessage } from "../format";

describe("format", () => {
  const testCases = [
    "lorem FOO",
    {},
    { tags: "lorem", command: "FOO" },
    { tags: {}, command: "FOO" },
    { tags: { lorem: true }, command: "FOO" },
    { tags: { lorem: "" }, command: "FOO" },
    { tags: { lorem: "lorem" } },
    { tags: { lorem: "lorem", ipsum: "ipsum" }, command: "FOO" },
    { tags: { lorem: "lorem\\ipsum dolor\rsit\namet;" }, command: "FOO" },
    { tags: { lorem: true }, command: "FOO" },
    { tags: { lorem: true }, prefix: { user: "user" }, command: "FOO" },
    { prefix: "lorem", command: "FOO" },
    { prefix: { name: "se.rv.er" }, command: "FOO" },
    { prefix: { user: "user" }, command: "FOO" },
    { prefix: { name: "name", user: "user" }, command: "FOO" },
    { prefix: { name: "name", host: "host" }, command: "FOO" },
    { prefix: { name: "name", user: "user", host: "host" }, command: "FOO" },
    { command: "FOO" },
    { command: "FOO", params: [] },
    { command: "FOO", params: ["lorem"] },
    { command: "FOO", params: ["lorem", "ipsum"] },
    { command: "FOO", params: ["lorem\\"] },
    { command: "FOO", params: [" dolor"] },
    { command: "FOO", params: ["\rsit"] },
    { command: "FOO", params: ["\namet;"] },
    { command: "FOO", middle: ["lorem", "ipsum"], trailing: "Lorem ipsum dolor sit amet" },
    { params: ["lorem", "ipsum"] },
  ];

  for (const testCase of testCases) {
    test(`should format "${JSON.stringify(testCase)}"`, () => {
      let result: string | FormatError;

      try {
        result = formatMessage(testCase as never);
      } catch (error) {
        result = error;
      }

      expect(result).toMatchSnapshot();
    });
  }
});
