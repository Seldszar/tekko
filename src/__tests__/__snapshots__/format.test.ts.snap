// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format should format ""lorem FOO"" 1`] = `"lorem FOO"`;

exports[`format should format "{"command":"FOO","middle":["lorem","ipsum"],"trailing":"Lorem ipsum dolor sit amet"}" 1`] = `"FOO lorem ipsum :Lorem ipsum dolor sit amet"`;

exports[`format should format "{"command":"FOO","params":[" dolor"]}" 1`] = `"FOO : dolor"`;

exports[`format should format "{"command":"FOO","params":["\\namet;"]}" 1`] = `
"FOO :
amet;"
`;

exports[`format should format "{"command":"FOO","params":["\\rsit"]}" 1`] = `
"FOO :
sit"
`;

exports[`format should format "{"command":"FOO","params":["lorem","ipsum"]}" 1`] = `"FOO lorem ipsum"`;

exports[`format should format "{"command":"FOO","params":["lorem"]}" 1`] = `"FOO lorem"`;

exports[`format should format "{"command":"FOO","params":["lorem\\\\"]}" 1`] = `"FOO :lorem\\"`;

exports[`format should format "{"command":"FOO","params":[]}" 1`] = `"FOO"`;

exports[`format should format "{"command":"FOO"}" 1`] = `"FOO"`;

exports[`format should format "{"params":["lorem","ipsum"]}" 1`] = `[FormatError: Invalid Message]`;

exports[`format should format "{"prefix":"lorem","command":"FOO"}" 1`] = `":lorem FOO"`;

exports[`format should format "{"prefix":{"name":"name","host":"host"},"command":"FOO"}" 1`] = `":name@host FOO"`;

exports[`format should format "{"prefix":{"name":"name","user":"user","host":"host"},"command":"FOO"}" 1`] = `":name!user@host FOO"`;

exports[`format should format "{"prefix":{"name":"name","user":"user"},"command":"FOO"}" 1`] = `":name FOO"`;

exports[`format should format "{"prefix":{"name":"se.rv.er"},"command":"FOO"}" 1`] = `":se.rv.er FOO"`;

exports[`format should format "{"prefix":{"user":"user"},"command":"FOO"}" 1`] = `[FormatError: Invalid Prefix]`;

exports[`format should format "{"tags":"lorem","command":"FOO"}" 1`] = `"@lorem FOO"`;

exports[`format should format "{"tags":{"lorem":""},"command":"FOO"}" 1`] = `"@lorem= FOO"`;

exports[`format should format "{"tags":{"lorem":"lorem","ipsum":"ipsum"},"command":"FOO"}" 1`] = `"@lorem=lorem;ipsum=ipsum FOO"`;

exports[`format should format "{"tags":{"lorem":"lorem"}}" 1`] = `[FormatError: Invalid Message]`;

exports[`format should format "{"tags":{"lorem":"lorem\\\\ipsum dolor\\rsit\\namet;"},"command":"FOO"}" 1`] = `"@lorem=lorem\\\\ipsum\\sdolor\\rsit\\namet\\: FOO"`;

exports[`format should format "{"tags":{"lorem":true},"command":"FOO"}" 1`] = `"@lorem FOO"`;

exports[`format should format "{"tags":{"lorem":true},"command":"FOO"}" 2`] = `"@lorem FOO"`;

exports[`format should format "{"tags":{"lorem":true},"prefix":{"user":"user"},"command":"FOO"}" 1`] = `[FormatError: Invalid Prefix]`;

exports[`format should format "{"tags":{},"command":"FOO"}" 1`] = `"FOO"`;

exports[`format should format "{}" 1`] = `[FormatError: Invalid Message]`;
